Multivariate scoring - WIP

nikosbosse opened the pull request:
<!-- Thanks for opening this pull request! Below we have provided a suggested template for PRs to this repository and a checklist to complete before opening a PR -->
 
## Description

This PR closes #<issue-number>.

[Describe the changes that you made in this pull request.]

## Checklist

- [ ] My PR is based on a package issue and I have explicitly linked it.
- [ ] I have included the target issue or issues in the PR title as follows: *issue-number*: PR title
- [ ] I have tested my changes locally.
- [ ] I have added or updated unit tests where necessary.
- [ ] I have updated the documentation if required.
- [ ] I have built the package locally and run rebuilt docs using roxygen2.
- [ ] My code follows the established coding standards and I have run `lintr::lint_package()` to check for style issues introduced by my changes. 
- [ ] I have added a news item linked to this PR.
- [ ] I have reviewed CI checks for this PR and addressed them as far as I am able.

<!-- Thanks again for this PR - @scoringutils dev team -->

sbfnk commented:
Wow, great to see this in action! I have just worked through the vignette and have two high-level comments:

#### 1. Defining the multivariate grouping

```{r}
example_multiv <- example_univ_multi

forecast_unit <- get_forecast_unit(example_multiv)
grouping <- setdiff(forecast_unit, c("location", "location_name"))
example_multiv <- set_grouping(example_multiv, grouping)
```

This seems a bit convoluted. Do you envision a situation where one would ever want to define the multivariate groupings via the columns that are common across the multivariate predictions, as opposed to ones that vary? Purely from looking at this example I would think it's more natural to work with option 1 in https://github.com/epiforecasts/scoringutils/issues/288#issuecomment-2523958177 i.e.

```{r}
example_multiv <- set_grouping(example_univ_multi, joint_across = c("location", "location_name"))
```

#### 2. Returning multivariate scores

I was surprised to see that `score(example_univ_multi)` still returns a data table with 4 rows. I think once you've defined the grouping it would be more natural to only calculate scores that we can do so jointly across the multivariate predictions, i.e. I think it would be more natural to return

```r
> score(example_multiv)
   target_end_date target_type forecast_date                 model horizon energy_score
            <Date>      <char>        <Date>                <char>   <num>        <num>
1:      2021-05-15       Cases    2021-05-03 EuroCOVIDhub-ensemble       2     54795.73
```

nikosbosse commented:
> This seems a bit convoluted. Do you envision a situation where one would ever want to define the multivariate groupings via the columns that are common across the multivariate predictions, as opposed to ones that vary?

I agree that grouping by the columns that vary is more natural. The reason I did it differently, for now, was consistency with how we set the forecast unit. But I guess it's fine to have a different approach here because we really group something

> I was surprised to see that score(example_univ_multi) still returns a data table with 4 rows.
Very open to changing that. We currently just do multivariate and univariate scores together. But we don't _have_ to return the univariate scores

sbfnk commented:
> But we don't _have_ to return the univariate scores

Could we not calculate them? This may also help with the performance issue. I guess it relates to https://github.com/epiforecasts/scoringutils/issues/288#issuecomment-2521027332 and comments around, i.e. do we want to have a separate set of metrics  / a different S3 class for multivariate scores or are they just different ways of computing scores. I'm beginning to see a case for treating them separately.

nikosbosse commented:
> Could we not calculate them? 
Yeah, we definitely don't have to calculate them. 

Funnily enough, calculating grouped multivariate forecasts is actually slightly faster than the non-grouped ones. I think the more expensive thing is calculating multivariate scores for univariate forecasts

seabbs commented:
> set_grouping(example_univ_multi, joint_across = c("location", "location_name"))

Still need to look at this but just swooping over to say the use of grouping, across and joint here makes me wonder if there is a better set of names i.e `set_joint` and then the arg is just across?

nikosbosse commented:
@seabbs Were you planning on doing some preliminary review of this, or should I just implement the different forecast class for this as discussed?

seabbs commented:
I'm very happy to do a full code review but unless there is a lot of pushback the different forecast class seems to be a much cleaner way to go to me

seabbs commented:
Let me know when I should look!

seabbs commented:
> do we want the grouping to be a "group by" or "group across" thing? Currently it's "group by" with a helper function to do "group across". We previously had the same discussion for summarise_scores.

Before I look at this my instinct is we follow exactly what we did with `summarise_scores` to save discussion unless there is a strong reason not to do so.

nikosbosse commented:
> Before I look at this my instinct is we follow exactly what we did with summarise_scores to save discussion unless there is a strong reason not to do so.

That's a good point. Basically, that's what I did here, but you'll realise that it's a bit clunky. 

Thinking about it, we might actually want to consolidate this "across" --> "group by" thing into a single helper function. One that you could use both for `summarise_by` and the multivariate grouping

seabbs commented on vignettes/scoring-multivariate-forecasts.Rmd:
```diff
@@ -0,0 +1,180 @@
+
+In the univariate case, Energy Score and CRPS are the same (see output above). Now, they are different, because one is treating the data as a multivariate forecast, the other as 4 separate univariate forecasts.
+
+If, at any point, you want to score the same forecast using different groupings, you'd have to specify those groupings separately and score the forecast multiple times.
```
maybe say create a new as_forecast object?

seabbs reviewed (approved):
I think this is really good and I just have minor comments. 

I think looking at this in detail it looks like across does make more sense as it seems really hard to set things by group? I know we had a lot of chat about this already but here it does seem different to that conversation?

My only other major comment is on naming but lets just ignore that unless it excites you

sbfnk commented on tests/testthat/test-class-forecast-sample-multivariate.R:
```diff
@@ -0,0 +1,186 @@
+# ==============================================================================
+# as_forecast_sample_multivariate()
+# ==============================================================================
+test_that("as_forecast_sample_multivariate() works as expected", {
```
this only tests that it doesn't error - do you also want a test that the results are correct and/or a snapshot test to check that future changes don't change them?

seabbs replied:
classic LLM test ;)

nikosbosse replied:
I (Claude) added a few snapshot tests. They are not super elegant as they need updating when we add new scores, but 🤷

sbfnk reviewed (commented):
This is really nice - just worked through the updated vignette and I think it's much improved.

What I don't like is that the "across" columns are still treated as defining a forecast unit:
```r
grouping <- define_grouping_cols(example_univ_multi, across = c("location", "location_name"))
example_multiv <- as_forecast_sample_multivariate(
  data = example_univ_multi,
  grouping = grouping
)
example_multiv
#> Forecast type: sample_multivariate
#> Forecast unit:
#> location, location_name, target_end_date, target_type, forecast_date, model, horizon, and
#> .scoringutils_group_id
```

Relatedly, in
```r
#> score(example_multiv)
#>    energy_score .scoringutils_group_id
#>           <num>                  <int>
#> 1:     54795.73                      1
```
the columns that do define the forecast unit got dropped. This means if I want to do this e.g. across multiple models I'll have to first reconstruct which grouping id corresponds to which model etc. which doesn't seem ideal?

nikosbosse commented on NEWS.md:
```diff
@@ -147,7 +147,7 @@ This major release contains a range of new features and bug fixes that have been
 - New `set_forecast_unit()` function allows manual setting of forecast unit.
-- `summarise_scores()` gains new `across` argument for summarizing across variables.
+- `summarise_scores()` gains new `across` argument for summarizing across variables. EDIT: This has since been removed again in [PR #831](https://github.com/epiforecasts/scoringutils/pull/831).
 - New `transform_forecasts()` and `log_shift()` functions allow forecast transformations. See the documentation for `transform_forecasts()` for more details and an example use case.
```
eh yeah we dropped it at some point and I just forgot to mention it back then... So I thought it would be helpful to at least have that documented somewhere

nikosbosse commented:
> the columns that do define the forecast unit got dropped. This means if I want to do this e.g. across multiple models I'll have to first reconstruct which grouping id corresponds to which model etc. which doesn't seem ideal?

oh, yeah, that's a bug I think... 

> What I don't like is that the "across" columns are still treated as defining a forecast unit, which I don't think they are any more:

You mean that they appear in the print statement? (and if you call `get_forecast_unit()`?). Yeah that makes sense, we can just hide it from there. 

>  especially as the grouping is really an internal thing and so ideally not really even seen my the user (or if they do see it they don't have to think about it)

My intuition is users would want to see it (so you can also e.g. reference issues in the group and also you still have the grouping column in the validated forecast object because that's how `score` knows about the group. Maybe we could call the column just `.group` though.

> I don't think we need to do this. My proposal would be to only allow the multivariate forecast group/across in the as_forecast_sample_multivariate transformer.

So we'd do `across` in the multivariate case and `by` in `summarise_scores()`?

seabbs commented:
> My intuition is users would want to see it (so you can also e.g. reference issues in the group and also you still have the

I think we shouldn't hide the internal variable but it shouldn't need to be used. It should be printed in the object.

nikosbosse commented:
> On a more general point is there a reason to use grouping rather than by in the current multivariate implementation

You mean as a name? Happy to us `by` 🤷

sbfnk commented:
So the things left to do are to 1) address the minor comments and 2) fix the bug around units of forecast?

nikosbosse commented on R/class-forecast-sample-multivariate.R:
```diff
@@ -0,0 +1,291 @@
+#' @export
+#' @keywords as_forecast transform
+# nolint start: object_name_linter
+as_forecast_sample_multivariate <- function(data, ...) {
```
Possibly :D No strong opinions. 
I'll go with it

nikosbosse commented on R/metrics-sample-multivariate.R:
```diff
@@ -0,0 +1,51 @@
+#' @inherit scoringRules::es_sample params
+#' @keywords internal_input_check
+#' @export
+energy_score_multivariate <- function(observed, predicted, grouping_id, w = NULL) {
```
Is that not clear from the function signature + docs here?

nikosbosse commented:
I think I addressed your comments: 
- added snapshot tests
- changed the name from `as_forecast_sample_multivariate()` to `as_forecast_multivariate_sample()`
- made it such that the output of `score()` now includes the grouping id variables, in addition to the `.scoringutils_grouping_id` column
- removed the helper function and added to the docs that you can use `setdiff` to get an `across` behaviour
- changed `grouping` to `by` in the function signature for `as_forecast_multivariate_sample()`

I note a small inconsistency where we now have the `.scoringutils_grouping_id` in the output for multivariate forecasts. We could do the same thing with the forecast unit, but are not doing it. 
Also, is `.grouping_id` better than `.scoringutils_grouping_id`? The former is nicer, the latter clearer and less ambiguous.

sbfnk commented:
@nikosbosse can you click on "resolve" on the comments that you've resolved? Are they all? Or are their open discussion points?

sbfnk reviewed (changes requested):
Looks great! A few minor niggles.

file class-forecast-multivariate-sample.R needs to be renamed

> I note a small inconsistency where we now have the .scoringutils_grouping_id in the output for multivariate forecasts.

Sorry, I'm not following. What do we have here that we don't have for the forecast unit?

> Also, is .grouping_id better than .scoringutils_grouping_id? The former is nicer, the latter clearer and less ambiguous.

I prefer the former

I think it would also be nice to print the grouping ids when printing a forecast object (as we print the forecast unit) but this is a nice to have, not an essential.

seabbs commented on R/class-forecast-sample-multivariate.R:
```diff
@@ -0,0 +1,279 @@
+  data[, .scoringutils_count := .N, by = eval(get_forecast_unit(data))]
+
+  for (group_id in unique(data$.scoringutils_group_id)) {
+    counts <- data[.scoringutils_group_id == group_id, .scoringutils_count]
```
Can we set a key and keep theirs?

nikosbosse commented:
> we do need to print the correct forecast unit:

hmmm. I see the point, @sbfnk. If we take it seriously, though, then the concept of `forecast_unit` and `by` only makes half sense. 

In the readme, we currently have: 

```
example_univ_multi <- example_sample_continuous[
  target_type == "Cases" &
    forecast_date == "2021-05-03" &
    target_end_date == "2021-05-15" &
    horizon == 2 &
    model == "EuroCOVIDhub-ensemble"
]


forecast_unit <- get_forecast_unit(example_univ_multi)
grouping <- setdiff(forecast_unit, c("location", "location_name"))

example_multiv <- as_forecast_multivariate_sample(
  data = example_univ_multi,
  by = grouping
)
```

`as_forecast_multivariate_sample()` also allows you to specify a `forecast_unit` argument. Ultimately, I think, the forecast unit is not a forecast unit, but instead a "unit of a single observation". For the univariate case that coincided with the forecast unit, but you're right to point out that the concept doesn't make much sense in the multivariate case. 

I'm slightly unsure what to do. We could keep the argument `forecast_unit` and document it and say that we mean the unit of a single observation? Give a new argument name for multivariate forecasts like `observation_unit`? Change it everywhere?

nikosbosse commented:
Latest update
- addresses the key setting issue by removing the key
- renamed forecast class file Seb mentioned
- asked difficult questions that might mean more updates

Left to do: 
- rename `.scoringutils_group_id` to `.group_id`
- make decision on forecast unit and observation unit

seabbs commented:
>  forecast unit, but instead a "unit of a single observation".

Agree

sbfnk commented:
> `as_forecast_multivariate_sample()` also allows you to specify a `forecast_unit` argument. Ultimately, I think, the forecast unit is not a forecast unit, but instead a "unit of a single observation". For the univariate case that coincided with the forecast unit, but you're right to point out that the concept doesn't make much sense in the multivariate case.

Just trying to think this through. We now have potentially:
1. a single numeric observation
1. a vector of numeric observations (length $n$) representing observations that are being forecast and should be evaluated jointly
1. a vector of predictions (length $m$) representing a probabilistic forecast of a single numeric observation (see (1)), e.g. quantiles or samples
1. a matrix of predictions (length $m \times n$) representing a probabilistic forecast of a vector of observations (see (2))

The reason we need to specify `by`, `forecast_unit` etc. is that we have a flat data structure so we need to identify which rows correspond to (2-4). At the moment we use `by` to define (2) and `forecast_unit` to define (3) - which then uniquely identifies (4). My confusion above stems from the fact that I interpreted the forecast unit to mean (2).

If we're in agreement up to this point then I think (for a future issue, I guess) it would make sense to define these somewhere in the docs, give them appropriate names and then use these names as the function arguments. The easiest / least breaking way might be to:
- continue defining (3) as the forecast unit
- give (2) another name (I think observation unit would make sense here actually but it's the reverse of what you, @nikosbosse suggest above so it might be confusing)
- rename `by` to whatever we agree on in the previous steps (as "the columns that define the xyz unit")
- document it all

nikosbosse commented:
Ah yes. I had similar thoughts (albeit less cleanly organised) when I tried to change the docs yesterday and realised that "model" is part of the forecast unit and so the forecast unit isn't actually the observation unit. 

What I ended up with was some notion of
"(univariate) forecast unit" and "multivariate forecast unit"  where 
- "forecast unit" defines the rows (samples, quantiles) that belong to a single forecast if we assumed that the forecast was a univariate one (regardless of whether that's the case or not)
- "multivariate forecast unit" defines the rows that belong to a single multivariate forecast

it kinda works with the existing framework, where you define the multivariate forecast using `forecast_unit` and `by`:
```
forecast_unit <- c("location", "target_end_date", "target_type", "forecast_date", "model", "horizon")
grouping <- setdiff(forecast_unit, c("location"))

example_multiv <- as_forecast_multivariate_sample(
  data = example_univ_multi,
  forecast_unit = forecast_unit, # this is the "univariate forecast unit"
  by = grouping
)
```
But it feels maybe a bit clunky. 
For a multivariate forecast, for example,`get_forecast_unit()` then has 2 different possible return values (only before scoring though, because after scoring the univariate forecast unit is no longer discernible - above, the location column would be gone).
On the other hand, it's close to what we have in the univariate case.

I think the basic insight is that univariate forecast unit, multivariate forecast unit, and locations grouped together are linked by `univariate_forecast_unit = c(multivariate_forecast_unit, group_together)`. (where currently we do `by = multivariate_forecast_unit = setdiff(forecast_unit, group_together)`.

In some sense, in the multivariate case, there is no "need" to specify `group_together` (insert better name) at all. You could just define the forecast unit directly and use that to group the data. The reason why it's helpful to do that specifically is that in the univariate case we can run a check that the univariate forecast unit is specified correctly via `get_duplicate_forecasts()`. We can't do that in the multivariate case because there will be duplicates by definition. Only if we specifically ask for the `forecast_unit` (or alternatively, `group_together`) are we able to run that check (+ do the automagical thing of kicking out irrelevant columns).

nikosbosse commented:
Sidenote: I'm starting to like the `.group_id` column and am wondering whether we should also have it in the univariate case as well? I think it could increase conceptual clarity and make it easier to issue warnings by saying "issue in group 4".

The most intuitive name that immediately comes to mind that would also work well for multivariate forecasts would be... "forecast_unit"🤦‍♂️. We need a UX designer at this point... 

<img width="819" height="368" alt="image" src="https://github.com/user-attachments/assets/a2b98c3c-157a-44a8-9072-7991990cbdf3" />

sbfnk commented:
My conclusion from all of this is that we could:
- define the forecast unit as the univariate forecast unit (thus, no changes needed)
- go back to using `joint_across` or maybe `multivariate_across` instead of `by` (noting that it's quite different to the `by` argument in `summarise_scores()` so this might help make the distinction, to define the multivariate forecast unit
- explain all of this in the docs
If you want (I don't feel strongly) we could then add 
- `.forecast_unit_id`
- `.multivariate_id`
or something like that.

seabbs commented:
My conclusion would be we: 

- use forecast_unit to mean observation unit and make plans to change the name
- in the univariate case this is the forecast unit
- in the multivariate case the forecast_unit is defined as a combination of the observation unit and the by variable

In terms of interface we could fairly easily add this by: 

- Depreciate forecast unit in favour of observation unit everywhere with warnings or if we wish allow both to be used with an error if ever specified and they differ.
- Here leave the forecat unit as missing and have it defined by by and observation unit. Check if forecat unit has been used and use that. Error if both forecast unit and observation unit have been used. 

The print methods etc would then still work as they do now?

seabbs commented:
I don't feel incredibly strongly though so happy to go with what you propose @nikosbosse

nikosbosse commented:
Nope, blocker is me. I'm currently unblocking myself. 

in @seabbs' proposal, the following isn't quite true: 
> in the univariate case this is the forecast unit
the forecast unit would include something like `model`, the observation unit not. 

For that reason I'm currently leaning towards @sbfnk's proposal to always ask the user to specify 
- the univariate forecast unit
- `joint_across`

Note: I _think_ we could also just have the user specify the multivariate forecast unit directly, without any `joint_across` variable. The effect of that would be
- the variable that we join across (e.g. the forecast horizon or the location) would get dropped
- we'd have to update `get_duplicate_forecasts()` to do something clever/convoluted like checking whether a single quantile is specified more often than we have observations

What I like about univariate forecast unit + `joint_across`:
- slightly easier to diagnose data issues, e.g. a wrongly specified forecast unit
- conceptual clarity: you start with a single forecast and then consciously evaluate some observations together. In your output, you still see all the columns (flipside of that is: you actually have to _have_ all the columns. I.e. you couldn't evaluate a multivariate forecast if you're not able to specify the univariate version, even if it's a completely valid multivariate forecast)
- you can evaluate different groupings without having to create a new forecast object every time (since columns don't get dropped, you just have to specify a different `joint_across`

nikosbosse commented:
I'm very confused... both the linter and the checks pass for me locally...

nikosbosse commented:
Recent updates: 
- Renamed `by` to `joint_across` as suggested by Seb, and updated the documentation for it. 
- renamed `group_id` to `mv_group_id`. But also not opposed to the `multivariate_id` Seb suggested. `mv_group_id` looks less nice, but is more descriptive
- fixed tests and linting issues

The PR should be ready for review now. It's a lot of file changes, but overall not that bad. I left other things for future PRs

nikosbosse commented:
Are folks happy with this as is? (my comments were just to make some of the code changes easier to follow and can all be resolved).

nikosbosse commented:
Yes, please!

seabbs commented on NEWS.md:
```diff
@@ -156,7 +156,7 @@ This major release contains a range of new features and bug fixes that have been
 
 - Documentation updated to reflect changes since version 1.1.0, including new transform and workflow functions.
```
No news item for this feature?

nikosbosse replied:
Good catch. Added a news item in my most recent commit

seabbs commented on vignettes/scoring-multivariate-forecasts.Rmd:
```diff
@@ -0,0 +1,168 @@
+
+If, at any point, you want to score the same forecast using different groupings, you'd have create a new separate forecast object with a different grouping and score that new forecast object.
+
+<!-- ## Details on the grouping - WIP
```
Is there an issue for this?

nikosbosse replied:
I don't think this warrants an issue - to me, this is jus the design choice and expected behaviour. 
We *could* have a function to specify a new `joint_across`, but to me this is just `as_forecast_multivariate_sample`

seabbs replied:
well if there is no future work planned, I don't see why the comment is still in here as WIP?

seabbs reviewed (approved):
This looks good I think. I wonder if people will be able to find the docs for the as_forecast_multivariate.dataframe method but that is perhaps no different to the other as_methods. I noted a few places for follow on issues and we also likely want some issues to write an adaptor from as_forecast_sample to as_forecast_multivariate. 

Nice!

nikosbosse commented on tests/testthat/_snaps/class-forecast-multivariate-sample.md:
```diff
@@ -0,0 +1,161 @@
+# as_forecast_multivariate_sample() creates expected structure
+
+    Code
```
#1043

nikosbosse commented:
@seabbs thank you for your review, I addressed your comments in the most recent commit
@sbfnk does this look good to merge to you?

seabbs commented:
Yes a customised method that uses the existing information in the sample object as you show

nikosbosse commented:
I might be missing something. But as far as I can tell, vanilla `as_forecast_multivariate_sample()` does all you could wish for
